File: Code_CS/C15_UserControls/CustomControls/CountedButton2.cs
using System;
using System.ComponentModel;
using System.Web.UI;
using System.Web.UI.WebControls;

namespace CustomControls
{
    [DefaultProperty("Text")]
    [ToolboxData("<{0}:CountedButton2 runat=server></{0}:CountedButton2>")]
    public class CountedButton2 : Button
    {
        private string displayString;

        // default constructor
        public CountedButton2()
        {
            displayString = "clicks";
            InitValues();
        }

        // overloaded, takes string to display (e.g., 5 books)
        public CountedButton2(string text)
        {
            displayString = text;
            InitValues();
        }

        // called by constructors
        private void InitValues()
        {
            if (ViewState["Count"] == null)
            {
                Count = 0;
            }
            Text = "Click me";
        }

        // count as property maintained in view state
        public int Count
        {
            get
            {
                return (int)ViewState["Count"];
            }

            set
            {
                ViewState["Count"] = value;
            }
        }

        // override the OnClick to increment the count,
        // update the button text and then invoke the base method
        protected override void OnClick(EventArgs e)
        {
            Count = Count + 1;
            Text = Count.ToString() + " " + displayString;
            base.OnClick(e);
        }
    }
}

File: Code_VB/C15_UserControls/CustomControls/CountedButton2.vb
Imports System 
Imports System.ComponentModel 
Imports System.Web.UI 
Imports System.Web.UI.WebControls 

Namespace CustomControls 
    <DefaultProperty("Text")> _ 
    <ToolboxData("<{0}:CountedButton2 runat=server></{0}:CountedButton2>")> _ 
    Public Class CountedButton2 
        Inherits Button 
        Private displayString As String 
        
        ' default constructor 
        Public Sub New() 
            displayString = "clicks" 
            InitValues() 
        End Sub 
        
        ' overloaded, takes string to display (e.g., 5 books) 
        Public Sub New(ByVal text As String) 
            displayString = text 
            InitValues() 
        End Sub 
        
        ' called by constructors 
        Private Sub InitValues() 
            If ViewState("Count") Is Nothing Then 
                Count = 0 
            End If 
            Text = "Click me" 
        End Sub 
        
        ' count as property maintained in view state 
        Public Property Count() As Integer 
            Get 
                Return CInt(ViewState("Count")) 
            End Get 
            
            Set(ByVal value As Integer) 
                ViewState("Count") = value 
            End Set 
        End Property 
        
        ' override the OnClick to increment the count, 
        ' update the button text and then invoke the base method 
        Protected Overloads Overrides Sub OnClick(ByVal e As EventArgs) 
            Count = Count + 1 
            Text = (Count.ToString() & " ") + displayString 
            MyBase.OnClick(e) 
        End Sub 
    End Class 
End Namespace 
